applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <caf/events/ble_common_event.h>
+
+#include "usb_event.h"
+
+#define MODULE dvfs
+#include <caf/events/module_state_event.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_DVFS_LOG_LEVEL);
+
+#include "ld_dvfs_handler.h"
+
+#define DVFS_RETRY_TIMEOUT		K_MSEC(500)
+/* Connection intervals used by LLPM are out of Bluetooth LE specification.
+ * The intervals are encoded by OR operation with a magic number of 0x0d00.
+ */
+#define REG_CONN_INTERVAL_LLPM_MASK	0x0d00
+
+enum dvfs_state {
+	DVFS_STATE_INITIALIZING,
+	DVFS_STATE_LLPM_CONNECTED,
+	DVFS_STATE_USB_CONNECTED,
+	DVFS_STATE_COUNT
+};
+
+/* Binary mask containing the frequency settings for each state.
+ * "1" means that the state requires the DVFS_FREQ_HIGH and "0" the DVFS_FREQ_MEDLOW.
+ * It is tailored for nrf54h20dk and might need to be adjusted for other platforms.
+ *
+ * DVFS_STATE_INITIALIZING	DVFS_FREQ_HIGH
+ * DVFS_STATE_LLPM_CONNECTED	DVFS_FREQ_MEDLOW
+ * DVFS_STATE_USB_CONNECTED	DVFS_FREQ_HIGH
+ */
+static uint8_t dvfs_freq_bitmask = 5;
+
+/* Binary mask tracking which states are requested.
+ * We start with DVFS_STATE_INITIALIZING on as it is active on start.
+ */
+static uint8_t dfvs_requests_state = 1;
+
+static enum dvfs_frequency_setting current_freq = DVFS_FREQ_HIGH;
+
+static struct dvfs_retry_work_data {
+	struct k_work_delayable dvfs_retry_work;
+	uint8_t dfvs_retries_cnt;
+} dvfs_retry_data;
+
+static const char *get_dvfs_frequency_setting_name(enum dvfs_frequency_setting setting)
+{
+	switch (setting) {
+	case DVFS_FREQ_HIGH: return "DVFS_FREQ_HIGH";
+	case DVFS_FREQ_MEDLOW: return "DVFS_FREQ_MEDLOW";
+	case DVFS_FREQ_LOW: return "DVFS_FREQ_LOW";
+	default: return "Unknown";
+	}
+}
+
+static const char *get_dvfs_state_name(enum dvfs_state state)
+{
+	switch (state) {
+	case DVFS_STATE_INITIALIZING: return "DVFS_STATE_INITIALIZING";
+	case DVFS_STATE_LLPM_CONNECTED: return "DVFS_STATE_LLPM_CONNECTED";
+	case DVFS_STATE_USB_CONNECTED: return "DVFS_STATE_USB_CONNECTED";
+	default: return "Unknown";
+	}
+}
+
+static void reschedule_dfvs_retry_work(enum dvfs_frequency_setting target_freq)
+{
+	if (dvfs_retry_data.dfvs_retries_cnt > 3) {
+		LOG_ERR("DVFS retry count exceeded.");
+		module_set_state(MODULE_STATE_ERROR);
+		return;
+	}
+	dvfs_retry_data.dfvs_retries_cnt++;
+	(void) k_work_reschedule(&dvfs_retry_data.dvfs_retry_work, DVFS_RETRY_TIMEOUT);
+}
+
+static void set_current_dvfs_frequency_and_reset_retries(enum dvfs_frequency_setting target_freq)
+{
+	LOG_INF("Have requested %s frequency", get_dvfs_frequency_setting_name(target_freq));
+	k_work_cancel_delayable(&dvfs_retry_data.dvfs_retry_work);
+	current_freq = target_freq;
+	dvfs_retry_data.dfvs_retries_cnt = 0;
+}
```

Comment:
This is not what i suggested in the https://github.com/nrfconnect/sdk-nrf/pull/16492#discussion_r1686251590 (suggested only the k_work_cancel_delayable + zeroing retries count).

I meant single responsibility, something like:
```
void cancel_dvfs_retry_work(void)
{
	k_work_cancel_delayable(&dvfs_retry_data.dvfs_retry_work);
	dvfs_retry_data.dfvs_retries_cnt = 0;
}
```

the frequency related stuff could have been like it was.